### ISSUE TITLE
Doxywizard Font Size Rendering Problem on 4k HiRes Monitors

### ISSUE THREAD

**Issue by cygnussystems** [doxygen/doxygen#10922]
I hadn't used DoxyWizard ever since I switched to a 4K high resolution monitor. I tried to work with doxywizard this morning but couldn't because of rendering issues. The application runs but renders with microscopic font size making it almost unusable. It's not just the font size but other aspects like tabs as well. This is a QT configuration issue! I had this problem in my own QT applications too.   There is a line of code that needs to be added to make QT applications run properly on high resolution monitors. 

Have a look at the screenshot - some of the text and tabs are microscopic. It's virtually impossible to read the generated program output. 

![doxygen_screenshot](https://github.com/doxygen/doxygen/assets/118726944/4da370fd-7df9-4e15-9e0d-4ddbb72905ed)

**Comment by albert-github** [comment]
- Which version of doxygen / doxywizard are you using?

You wrote:
> This is a QT configuration issue! I had this problem in my own QT applications too. There is a line of code that needs to be added to make QT applications run properly on high resolution monitors.

so: it would be useful when you added this to this issue so we don't have to invent the wheel again.

I think this problem is related to: https://stackoverflow.com/a/78570272/1657886 and here there is already a similar remark.

**Comment by cygnussystems** [comment]
I am  using the latest version available from the doxygen download web site. It is 1.11.0

The StackExchange comment was also written by myself this morning:-) That is when I had the bright idea to post it here too. We have an inhouse QT development that ran into the same  problem. I remember it was just a couple lines of code that needed to be added but I also remember that the first fix didn't work. I need to ask the person who managed to address the issue just what exactly they did to fix it. I will post again when I hear back.

**Comment by albert-github** [comment]
In the mean time I found the stackoverflow comment: https://stackoverflow.com/a/48582594/1657886 and here are 2 suggestions:
- setting environment variables and setting `QT_SCALE_FACTOR` seems to have an effect (as on my, non 4K, display I see larger characters when setting this variable to `2` and a bigger generated window). Also other environment variables might have an effect.
- adding some code:
  ```
  static const char ENV_VAR_QT_DEVICE_PIXEL_RATIO[] = "QT_DEVICE_PIXEL_RATIO";
  if (!qEnvironmentVariableIsSet(ENV_VAR_QT_DEVICE_PIXEL_RATIO)
        && !qEnvironmentVariableIsSet("QT_AUTO_SCREEN_SCALE_FACTOR")
        && !qEnvironmentVariableIsSet("QT_SCALE_FACTOR")
        && !qEnvironmentVariableIsSet("QT_SCREEN_SCALE_FACTORS")) {
    QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
  }
  ```

I've just pushed a proposed patch, pull request #10923
Unfortunately I cannot test it myself so it would be highly appreciated when somebody else could test the version without using environment variables.

Note: 
An executable can be obtained from the artifacts with the GitHub Actions: https://github.com/doxygen/doxygen/actions/runs/9353309813?pr=10923 this will be a dynamically linked executable so one needs also in the runtime environment the right dlls. Fot the subject of the dlls see https://github.com/doxygen/doxygen/issues/10917#issuecomment-2143863114

**Comment by cygnussystems** [comment]
ok, so this is some code that I got back from my colleague who worked on our QT application:

# Enable dpi scale
QApplication.setHighDpiScaleFactorRoundingPolicy(
Qt.HighDpiScaleFactorRoundingPolicy.PassThrough)
QApplication.setAttribute(Qt.AA_EnableHighDpiScaling, True)
QApplication.setAttribute(Qt.AA_UseHighDpiPixmaps, True)
# Query DPI Awareness (Windows 10 and 8)
awareness = ctypes.c_int()
errorCode = ctypes.windll.shcore.GetProcessDpiAwareness(0, ctypes.byref(awareness))

# Set DPI Awareness (Windows 10 and 8)
error_code = ctypes.windll.shcore.SetProcessDpiAwareness(2)
# the argument is the awareness level, which can be 0, 1 or 2

QApplication.setDesktopSettingsAware(True)
# os.environ["QT_AUTO_SCREEN_SCALE_FACTOR"] = "1"
# os.environ["QT_ENABLE_HIGHDPI_SCALING"] = "1"
# Finally create QApplication
q_application = QApplication(sys.argv)

 My monitor is 32 inch LG  32UN88A - and I run Windows with Scale 200% - all windows applications look normal, including our own QT application using the code above. Hopefully this will fix Doxywizard too!  I'm happy to test any new doxywizard.exe on my system.

**Comment by albert-github** [comment]
Thanks for the code.

I think there are already 2 things you can do
- test the proposed patch like I wrote in https://github.com/doxygen/doxygen/issues/10922#issuecomment-2145581039
- try the current 1.11.0 doxygen / doxywizard version with `set QT_SCALE_FACTOR=2` in a terminal window.

Back to the provided code (not just to comment the code but also to give some thoughts / things I found and to give some side information for other developers):
1. `QApplication.setHighDpiScaleFactorRoundingPolicy(Qt.HighDpiScaleFactorRoundingPolicy.PassThrough)` it looks like this is the default see e.g. https://doc.qt.io/qt-5/qguiapplication.html#setHighDpiScaleFactorRoundingPolicy
2.  `QApplication.setAttribute(Qt.AA_EnableHighDpiScaling, True)` this is present in the proposed patch
3.  `QApplication.setAttribute(Qt.AA_UseHighDpiPixmaps, True)` see https://doc.qt.io/qt-5/qt.html#ApplicationAttribute-enum we have to see whether or not setting this attribute is necessary based on tests
4.  `error_code = ctypes.windll.shcore.SetProcessDpiAwareness(2)` this is not a Qt variable but it looks like a typical windows setting, see https://learn.microsoft.com/en-us/windows/win32/api/shellscalingapi/nf-shellscalingapi-setprocessdpiawareness and https://learn.microsoft.com/en-us/windows/win32/api/shellscalingapi/ne-shellscalingapi-process_dpi_awareness . Looks like a sensible setting for windows.
5. `QApplication.setDesktopSettingsAware(True)` it looks like this is the default see e.g. https://doc.qt.io/qt-5/qguiapplication.html#setDesktopSettingsAware

@doxygen can you shine a light on this issue as well?

In my previous comment I forgot to mention the article https://doc.qt.io/qt-5/highdpi.html

I also found in https://doc.qt.io/qtforpython-6.2/porting_from2.html:
> The High DPI scaling attributes Qt.AA_EnableHighDpiScaling, Qt.AA_DisableHighDpiScaling and Qt.AA_UseHighDpiPixmaps are deprecated. High DPI is by default enabled in Qt 6 and cannot be turned off.

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.12.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
